ADR7 regarding embedding Java and version support (#29624)
gradle/gradle: architecture/standards/0007-java-pre-requisite.md

> # ADR-0007 - Gradle and Java as a pre-requisite

Reviewer: 🤔 I don't think this title is very clear.  As a pre-requisite to what?  Is Gradle and Java a pre-requisite for something?  Or is this talking about Gradle, and how Gradle uses Java as a pre-requisite for something?

> # ADR-0007 - Gradle and Java as a pre-requisite

Reviewer: ```suggestion
# ADR-0007 - Gradle and Java as a prerequisite
```

Author: Reworked the title, no longer applicable

## Date

2024-06-26

## Context

There have been discussions on embedding a Java runtime in the Gradle distribution.
This would allow users to run Gradle without having to install a Java runtime beforehand.

> Which is something that can ease Gradle adoption outside of the JVM ecosystem.

Reviewer: ```suggestion
This capability could help increase Gradle adoption outside of the JVM ecosystem.
```

> Gradle has different runtime components:

Reviewer: 💭 Maybe this section can be restructured to be more clear about there being 3 different ways to invoke Gradle (the wrapper, the tooling API, or running the `gradle` launcher directly), each of which will result in 3 different types of component processes spawned from the Gradle distribution.

Author: Added something

* The [Gradle Wrapper](https://docs.gradle.org/8.8/userguide/gradle_wrapper.html)

> * The [Gradle Distribution](https://gradle.org/releases/) which contains

Reviewer: ```suggestion
* The [Gradle Distribution](https://gradle.org/releases/) which will spawn various processes:
```

  * the [Gradle Launcher](https://blog.gradle.org/how-gradle-works-1#local-gradle-distribution-in-cli)
  * the [Gradle Daemon](https://docs.gradle.org/8.8/userguide/gradle_daemon.html)
  * the [Gradle Worker processes](https://docs.gradle.org/8.8/userguide/worker_api.html)
* The [Gradle Tooling API client](https://docs.gradle.org/8.8/userguide/third_party_integration.html#embedding)

At the moment, each of those components require a Java runtime to run.

> The recommended way of integrating a project with Gradle is to use the Gradle Wrapper.

Reviewer: ```suggestion
The recommended way of invoking Gradle is to use the Gradle Wrapper.
```

> The Wrapper then downloads the Gradle distribution, starts the Launcher and runs the build inside the Daemon.

Reviewer: ```suggestion
The Wrapper then downloads the Gradle distribution (if necessary), starts the Launcher and runs the build, spawning and connecting to other processes such as the Daemon as required.
```

> Embedding a Java runtime in the distribution would provide benefits.
> The Launcher, Daemon and Workers could run on it, removing the pre-requisite of an installed Java runtime.

Reviewer: ```suggestion
Embedding a Java runtime in the distribution would provide some benefits, such as allowing the Launcher, Daemon and Workers to run on it, removing the prerequisite of an installed Java runtime.
```

> However, this does not fully remove the pre-requisite, as the Wrapper would still need a Java runtime.

Reviewer: ```suggestion
However, this does not fully remove the prerequisite, as the Wrapper itself would still need an installed Java runtime to execute.
```

> Gradle needs a solution that covers the Wrapper use-case as well given its importance.

Reviewer: ```suggestion
Gradle needs a solution that covers the use-case of starting the Wrapper itself as well given its importance.
```

> In addition, some of these discussions included conversation about having a single Java version supported by the Gradle Launcher and Daemon.

Reviewer: ```suggestion
In addition, some of these discussions included proposals for having a single Java version supported by the Gradle Launcher and Daemon.
```

> However, this would limit the ability of the Gradle ecosystem of plugin authors to take advantage of new Java features and improvements.

Reviewer: ```suggestion
However, this would limit the ability of the Gradle ecosystem plugin authors to take advantage of new Java features and improvements.
```

> One example is when a popular plugin started requiring Java 17 because the underlying framework behind the plugin did and it is used in the plugin itself.

Reviewer: ```suggestion
One example is when a popular plugin started requiring Java 17 because the underlying framework behind the plugin began requiring 17, and code from the framework was used in the plugin itself.
```

What would have happened if Gradle was limited to Java 11 at the time?

## Decision

1. The Gradle distribution will never include a Java runtime.
   Instead, Gradle will leverage Java toolchains for the Daemon and worker processes.

> 2. To stop requiring a pre-installed Java runtime for the Gradle Wrapper and Launcher, Gradle will, in the future, develop a native version of those.

Reviewer: ```suggestion
2. In the future, to stop requiring a pre-installed Java runtime for the Gradle Wrapper and Launcher, Gradle will develop native versions of those components.
```

> 2. To stop requiring a pre-installed Java runtime for the Gradle Wrapper and Launcher, Gradle will, in the future, develop a native version of those.

Reviewer 2: ❓ This might take a long while until we get there. Does this decision prevent us to implement an intermediary solution earlier? IOW, can this decision be interpreted as not mandating THE solution is native and nothing else should be done until we have native launchers? I would like this to be clarified.

Reviewer 2: Also, I think future things are a better fit for the `Consequences` section than for the `Decision` one.

Author: I did move the point and clarified that there is no decided or rejected solution for this.

> 3. The Gradle Daemon, Worker processes, and Tooling API client will support running on more than one Java version.

Reviewer: ```suggestion
3. The Gradle Daemon, Worker processes, and Tooling API client will continue to support running on different Java versions during a single Gradle invocation.
```
